webpack/webpack-embedded-dependencies-plugin/src/EmbeddedDependenciesWebpackPlugin.ts

```diff
@@ -0,0 +1,384 @@
+      const thirdPartyPackages: ThirdPartyPackageMap = new Map();
+
+      normalModuleFactory.hooks.module.tap(
+        PLUGIN_NAME,
+        (module, moduleCreateData: IWebpackModuleCreateData, resolveData) => {
+          const { resourceResolveData } = moduleCreateData;
+          const pkg: IPackageData | undefined = resourceResolveData?.descriptionFileData;
+          const filePath: string | undefined = resourceResolveData?.descriptionFileRoot;
+
+          if (
+            pkg &&
+            filePath &&
+            this._packageFilterFunction(pkg, filePath) &&
+            filePath?.includes('node_modules')
+          ) {
+            const key: PackageNameAndVersion = makePackageMapKeyForPackage(pkg);
+            thirdPartyPackages.set(key, { packageFolderPath: filePath, packageJsonData: pkg });
+          }
+
+          return module;
+        }
+      );
+
+      compilation.hooks.processAssets.tapPromise(
+        { name: PLUGIN_NAME, stage: Compilation.PROCESS_ASSETS_STAGE_REPORT },
+        async (assets) => {
+          const packages: IPackageData[] = [];
+
+          try {
+            await Async.forEachAsync(
+              thirdPartyPackages,
+              async ([, { packageFolderPath: dir, packageJsonData: data }]) => {
+                const { name, version } = data;
+                let licenseSource: string | undefined;
+                const license: string | undefined = parseLicense(data);
+                const licensePath: string | undefined = await this._getLicenseFilePathAsync(dir, compiler);
+                if (licensePath) {
+                  licenseSource = await FileSystem.readFileAsync(licensePath);
+
+                  const copyright: string | undefined =
+                    this._parseCopyright(licenseSource) || parsePackageAuthor(data);
+
+                  packages.push({
+                    name,
+                    version,
+                    license,
+                    licenseSource,
+                    copyright
+                  });
+                } else {
+                  // If there is no license file path, we still should populate the other required fields
+                  const copyright: string | undefined = parsePackageAuthor(data);
+
+                  packages.push({
+                    name,
+                    version,
+                    license,
+                    copyright
+                  });
+                }
+              }
+            );
+          } catch (error) {
+            this._emitWebpackError(compilation, 'Failed to process embedded dependencies', error);
+          } finally {
+            Sort.sortBy(packages, (pkg) => pkg.name);
+          }
+
+          const dataToStringify: IEmbeddedDependenciesFile = {
+            embeddedDependencies: packages
+          };
+
+          compilation.emitAsset(this._outputFileName, new sources.RawSource(JSON.stringify(dataToStringify)));
```

Comment:
`sources` should come from `compiler.webpack.sources` to guarantee version synchronization. Use type-only imports for everything from webpack.